src/Microsoft.Diagnostics.Monitoring.RestServer/DiagProcessFilter.cs

```diff
@@ -0,0 +1,153 @@
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Monitoring.RestServer
+{
+    /// <summary>
+    /// There are effectively 3 sets of models for this:
+    /// The configuration model (ProcessFilterOptions), which does not contain criteria such as RuntimeId, and is always string based.
+    /// The REST call model (ProcessKey), which does not contain criteria such as Command Line, and is strongly typed.
+    /// This internal model, which contains all possible criteria and is string based. Only this model is used for the actual comparison.
+    /// </summary>
+    internal sealed class DiagProcessFilter
+    {
+        public IList<DiagProcessFilterEntry> Filters { get; set; } = new List<DiagProcessFilterEntry>();
+
+        public static DiagProcessFilter FromProcessKey(ProcessKey processKey)
+        {
+            var filter = new DiagProcessFilter();
+            filter.Filters.Add(TransformKey(processKey));
+            return filter;
+        }
+
+        public static DiagProcessFilter FromConfiguration(ProcessFilterOptions options)
+        {
+            var filter = new DiagProcessFilter();
+            foreach(ProcessFilterDescriptor processFilter in options.Filters)
+            {
+                filter.Filters.Add(TransformDescriptor(processFilter));
+            }
+            return filter;
+        }
+
+        private static DiagProcessFilterEntry TransformKey(ProcessKey processKey)
+        {
+            if (processKey.ProcessId.HasValue)
+            {
+                return new DiagProcessFilterEntry { Criteria = DiagProcessFilterCriteria.ProcessId, MatchType = DiagProcessFilterMatchType.Exact, Value = processKey.ProcessId.Value.ToString(CultureInfo.InvariantCulture) };
+            }
+            if (processKey.ProcessName != null)
+            {
+                return new DiagProcessFilterEntry { Criteria = DiagProcessFilterCriteria.ProcessName, MatchType = DiagProcessFilterMatchType.Exact, Value = processKey.ProcessName };
+            }
+            if (processKey.RuntimeInstanceCookie.HasValue)
+            {
+                return new DiagProcessFilterEntry { Criteria = DiagProcessFilterCriteria.RuntimeId, MatchType = DiagProcessFilterMatchType.Exact, Value = processKey.RuntimeInstanceCookie.Value.ToString("D") };
+            }
+
+            throw new ArgumentException($"Invalid {nameof(processKey)}");
+        }
+
+        private static DiagProcessFilterEntry TransformDescriptor(ProcessFilterDescriptor processFilterDescriptor)
+        {
+            switch (processFilterDescriptor.Key)
+            {
+                case ProcessFilterKey.ProcessId:
+                    return new DiagProcessFilterEntry { Criteria = DiagProcessFilterCriteria.ProcessId, MatchType = DiagProcessFilterMatchType.Exact, Value = processFilterDescriptor.Value };
```

Comment:
I think if we are going to only allow exact match (which makes sense), we need to validate that the ProcessFilterDescriptor does the same for ProcessId (not fail the check, but maybe write a warning that states it's going to only use Exact).

Reply:
I will defer this for now. I would rather do a broader set of ValidationAttribute/ValidationContext for all the config.